<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Première-solution-(naïve)-:-incorrect,-mais-complète-et-qui-termine." data-toc-modified-id="Première-solution-(naïve)-:-incorrect,-mais-complète-et-qui-termine.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Première solution (naïve) : incorrect, mais complète et qui termine.</a></span></li><li><span><a href="#Solution-correcte,-complète-et-qui-termine" data-toc-modified-id="Solution-correcte,-complète-et-qui-termine-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Solution correcte, complète et qui termine</a></span></li><li><span><a href="#Une-solution-qui-ne-termine-pas" data-toc-modified-id="Une-solution-qui-ne-termine-pas-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Une solution qui ne termine pas</a></span></li><li><span><a href="#Une-solution-incomplète-mais-correcte-et-qui-termine" data-toc-modified-id="Une-solution-incomplète-mais-correcte-et-qui-termine-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Une solution incomplète mais correcte et qui termine</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></div>

# Complétude, Correction, Terminaison.

Pour reprendre ce qui est écrit sur Wikipédia :
> Ces trois notions « correction », « complétude », « terminaison » sont liées, 
> et supposent qu'un algorithme est écrit pour résoudre un problème.
>
> La terminaison est l'assurance que l'algorithme terminera en un temps fini. Les preuves de terminaison font 
> habituellement intervenir une fonction entière positive strictement décroissante à chaque « pas » de l'algorithme.
>
> Étant donnée la garantie qu'un algorithme terminera, la preuve de correction doit apporter l'assurance 
> que si l'algorithme termine en donnant un résultat, alors ce résultat est effectivement une solution 
> au problème posé. Les preuves de correction font habituellement intervenir une spécification logique 
> que doivent vérifier les solutions du problème. La preuve de correction consiste donc à montrer 
> que les résultats de l'algorithme vérifient cette spécification.

> La preuve de complétude garantit que, pour un espace de problèmes donné, l'algorithme, 
> s'il termine, donnera l'ensemble des solutions de l'espace du problème. 
> Les preuves de complétude demandent à identifier l'espace du problème et 
> l'espace des solutions pour ensuite montrer que l'algorithme produit 
> bien le second à partir du premier. 

Prenons donc un problème (simple), résoudre une équation paramétrique diophantienne (recherche des solutions entières seulement, parce que l'informatique cela ne calcule bien et facilement qu'avec les entiers) :
$$ 2X+P = 0 $$

Remarques : l'espace des solutions est l'espace des entiers, au paramètre $-2N$ correspond la solution $N$. Pour les paramètres impairs, il n'y a pas de solution (entière).

Et proposons quelques programmes pour résoudre ce problème. Et pour chacun, regardons :
* est-ce que le programme, lorsqu'il donne un résultat, donne bien un résultat entier qui valide l'équation ; c'est à dire, est-ce que le programme est correct ?
* est-ce que le programme est complet ? 
* est-ce que le programme termine (toujours) ?

## Première solution (naïve) : incorrect, mais complète et qui termine.

In [25]:
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char * argv[]) {
int Parametre=10;
int Solution = -Parametre/2;
printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);
return 0;}

Solution de l'equation 2X+10 = 0 :
   X = -5
Verification 2*-5+10 = 0


Cela semble pas mal, mais :
<div style="width: 400px;  padding-top:10px; padding-bottom:10px;border: 3px solid #ff5252; background: #fce6d4;"> 
  <ul>
    <li><b>Non</b>, le programme n'est pas correct</li>
    <li><b>Oui</b>, le programme est complet</li>
    <li><b>Oui</b>, le programme  termine  toujours</li>
  </ul>
</div>

Pour voir le problème de correction, prenons un paramètre impair, par exemple P=9 

In [26]:
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char * argv[]) {
int Parametre=9;
int Solution = -Parametre/2;
printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);
return 0;}

Solution de l'equation 2X+9 = 0 :
   X = -4
Verification 2*-4+9 = 1


## Solution correcte, complète et qui termine

Pour obtenir la correction, il suffit de ne pas dire de bêtise (pour les paramètres impairs, il n'y a pas de solution) : 

In [28]:
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char * argv[]) {
int Parametre=9;
int Solution = -Parametre/2;
if (Parametre&1) {
  printf("Pas de solution entiere pour l'equation 2X+%d = 0 (%d est impair)\n",Parametre,Parametre);}
else {
  printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);}
return 0;}

Pas de solution pour l'equation 2X+9 = 0 (9 est impair)


Pour les paramètres pairs, cela reste correct.

In [29]:
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char * argv[]) {
int Parametre=10;
int Solution = -Parametre/2;
if (Parametre&1) {
  printf("Pas de solution entiere pour l'equation 2X+%d = 0 (%d est impair)\n",Parametre,Parametre);}
else {
  printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);}
return 0;}

Solution de l'equation 2X+10 = 0 :
   X = -5
Verification 2*-5+10 = 0


Voila donc un programme qui parait idéal :
<div style="width: 400px;  padding-top:10px; padding-bottom:10px;border: 3px solid #5eff52;background: #d4fcd4;"> 
  <ul>
    <li><b>Oui</b>, le programme est correct</li>
    <li><b>Oui</b>, le programme est complet</li>
    <li><b>Oui</b>, le programme termine toujours</li>
  </ul>
</div>
<pre>
(en fait, pour les grands entiers, il y aurait des choses à dire, voir plus loin)
(mais pour faire "simple", on dira que ce programme est ok)
(pour faire "simple" et vraiment correct, passer à Python (?))
</pre>

## Une solution qui ne termine pas

Je ne dis pas que le programme qui suit est idéal ... c'est juste un programme qui essaie de trouver un entier qui répond à l'équation !

De temps en temps cela marche :

In [33]:
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char * argv[]) {
int Parametre=-10;
int Solution;
for(Solution=0;2*Solution+Parametre!=0;Solution++) ;
printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);
return 0;}

Solution de l'equation 2X+-10 = 0 :
   X = 5
Verification 2*5+-10 = 0


De temps en temps, cela donne des résultats étranges (pour comprendre, il faudrait un petit cours sur les nombres en machine)

In [43]:
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char * argv[]) {
int Parametre=10;
int Solution;
for(Solution=0;2*Solution+Parametre!=0;Solution++) ;
printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);
return 0;}

Solution de l'equation 2X+10 = 0 :
   X = 2147483643
Verification 2*2147483643+10 = 0


Est-ce que c'est correct ? on dirait, vu la vérification, mais cela reste étrange. (l'explication est peut-être plus facile quand on regarde comment s'écrit 2147483643 en hexadécimal : 0x7FFFFFFB)

Est-ce complet ? il semble qu'il y ait une solution plus simple qui n'est pas donnée ?
(pour avoir un programme un peu plus complet on aurait pu mettre : ... )

In [12]:
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char * argv[]) {
int Parametre=10;
int Solution=0;
if (Parametre==0) {
  printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);}
for(Solution=1;Solution;Solution++) {
  if (2*Solution+Parametre==0) {
    printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);}}
return 0;}

Solution entiere de l'equation 2X+10 = 0 :
   X = 2147483643
Verification 2*2147483643+10 = 0

Solution entiere de l'equation 2X+10 = 0 :
   X = -5
Verification 2*-5+10 = 0



Pour la correction et la complétude ce n'est pas clair, mais il y a pire, pour la terminaison avec un paramètre impair on a un problème (déjà les calculs étaient longs, mais là !) :

In [ ]:
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char * argv[]) {
int Parametre=9;
int Solution;
for(Solution=0;2*Solution+Parametre!=0;Solution++) ;
printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);
return 0;}

Pas de réponse au bout de 5 minutes, vous pensez qu'il faut attendre encore un peu ?

On a donc là un programme :
<div style="width: 400px;  padding-top:10px; padding-bottom:10px;border: 3px solid #ff5252;background: #fce6d4;"> 
  <ul>
    <li><b>Oui/Non</b>, qui n'est pas clairement correct</li>
    <li><b>Non</b>, qui n'est pas complet</li>
    <li><b>Non</b>, qui ne termine pas toujours (parfois seulement)</li>
  </ul>
</div>


## Une solution incomplète mais correcte et qui termine 

En reprenant le programme précédent, même s'il n'est pas très efficace, on peut obtenir quelque chose de mieux mais limité aux "petits" entiers (de toute façon en machine, c'est toujours de petits entiers, sauf à faire de l'évaluation en précision infinie, mais c'est une autre histoire ...)

Le programme termine pour les paramètres impairs :

In [11]:
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char * argv[]) {
int Parametre=9;
int Solution;
if (Parametre==0) {
  printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);}
for(Solution=1;Solution<1000000000;Solution++) {
  if (2*Solution+Parametre==0) {
    printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);}
  if (-2*Solution+Parametre==0) {
    printf("Solution de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,-Solution,-Solution,Parametre,-2*Solution+Parametre);}}
printf("Fin de la recherche des solutions entieres de l'equation 2X+%d = 0\n",Parametre);
return 0;}

Fin de la recherche des solutions entieres de l'equation 2X+9 = 0


Le programme donne des solutions pour les "petits" paramètres pairs :

In [10]:
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char * argv[]) {
int Parametre=10;
int Solution;
if (Parametre==0) {
  printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);}
for(Solution=1;Solution<1000000000;Solution++) {
  if (2*Solution+Parametre==0) {
    printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);}
  if (-2*Solution+Parametre==0) {
    printf("Solution de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,-Solution,-Solution,Parametre,-2*Solution+Parametre);}}
printf("Fin de la recherche des solutions entieres de l'equation 2X+%d = 0\n",Parametre);
return 0;}

Solution de l'equation 2X+10 = 0 :
   X = -5
Verification 2*-5+10 = 0
Fin de la recherche des solutions entieres de l'equation 2X+10 = 0


Mais pour les gros paramètres, il ne donne pas de solution (il termine pourtant) :

In [9]:
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char * argv[]) {
int Parametre=2000000000;
int Solution;
if (Parametre==0) {
  printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);}
for(Solution=1;Solution<1000000000;Solution++) {
  if (2*Solution+Parametre==0) {
    printf("Solution entiere de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,Solution,Solution,Parametre,2*Solution+Parametre);}
  if (-2*Solution+Parametre==0) {
    printf("Solution de l'equation 2X+%d = 0 :\n   X = %d\nVerification 2*%d+%d = %d\n",Parametre,-Solution,-Solution,Parametre,-2*Solution+Parametre);}}
printf("Fin de la recherche des solutions entieres de l'equation 2X+%d = 0\n",Parametre);
return 0;}

Fin de la recherche des solutions entieres de l'equation 2X+2000000000 = 0


on a donc là :
<div style="width: 400px;  padding-top:10px; padding-bottom:10px;border: 3px solid #ff5252;background: #fce6d4;"> 
  <ul>
    <li><b>Oui</b>, un programme correct</li>
    <li><b>Non</b>, un programme incomplet</li>
    <li><b>Oui</b>, un programme qui termine toujours</li>
  </ul>
</div>

(mais un programme pas efficace !)

## Conclusion

Il y a donc des programmes de tous les genres !

(on peut donner des exemples pour les 8 cas de figure, ici, il n'y en a que 4+1, les autres s'en déduisent)
